src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -0,0 +1,474 @@
+            {
+                _diagnosticIds = null;
+            }
+
+            public IDELatestDiagnosticGetter(DiagnosticIncrementalAnalyzer owner, ImmutableHashSet<string> diagnosticIds, Solution solution, ProjectId projectId, bool includeSuppressedDiagnostics) :
+                this(owner, diagnosticIds, solution, projectId, documentId: null, includeSuppressedDiagnostics: includeSuppressedDiagnostics)
+            {
+            }
+
+            public IDELatestDiagnosticGetter(DiagnosticIncrementalAnalyzer owner, ImmutableHashSet<string> diagnosticIds, Solution solution, ProjectId projectId, DocumentId documentId, bool includeSuppressedDiagnostics) :
+                base(owner, solution, projectId, documentId, id: null, includeSuppressedDiagnostics: includeSuppressedDiagnostics)
+            {
+                _diagnosticIds = diagnosticIds;
+            }
+
+            public async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(CancellationToken cancellationToken)
+            {
+                if (CurrentSolution == null)
+                {
+                    return GetDiagnosticData();
+                }
+
+                DocumentId nullTargetDocumentId = null;
+
+                if (CurrentProjectId != null)
+                {
+                    await AppendDiagnosticsAsync(CurrentProject, nullTargetDocumentId, SpecializedCollections.EmptyEnumerable<DocumentId>(), cancellationToken).ConfigureAwait(false);
+                    return GetDiagnosticData();
+                }
+
+                await AppendDiagnosticsAsync(CurrentSolution, cancellationToken).ConfigureAwait(false);
+                return GetDiagnosticData();
+            }
+
+            protected override bool ShouldIncludeDiagnostic(DiagnosticData diagnostic)
+            {
+                return _diagnosticIds == null || _diagnosticIds.Contains(diagnostic.Id);
+            }
+
+            protected override async Task AppendDiagnosticsAsync(Project project, DocumentId targetDocumentId, IEnumerable<DocumentId> documentIds, CancellationToken cancellationToken)
+            {
+                // when we return cached result, make sure we at least return something that exist in current solution
+                if (project == null)
+                {
+                    return;
+                }
+
+                // get analyzers that are not suppressed.
+                // REVIEW: IsAnalyzerSuppressed call seems can be quite expensive in certain condition. is there any other way to do this?
+                var stateSets = StateManager.GetOrCreateStateSets(project).Where(s => ShouldIncludeStateSet(project, s)).ToImmutableArrayOrEmpty();
+
+                var analyzerDriverOpt = await Owner._compilationManager.CreateAnalyzerDriverAsync(project, stateSets, IncludeSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
```

Comment:
That will help only for FixAll in Solution cases. It won't help when:

1. FixAll in Project
2. Existing code fixers with their own FixAllProvider which compute diagnostics for one project at a time even for solution level fix all - it will become sequential computation for them.